### ISSUE TITLE
ArchLinux32 Support

### ISSUE THREAD

**Issue by Hayao0819** [BrenekH/blinky#21]
I have an project written in Golang which is for building pacman repository, which is [Hayao0819/ayaka](https://github.com/Hayao0819/ayaka)

Now, ayaka has no functions to update db and upload it but has building package with devtools so your blinky has what my tools does not have.

What I would like to do is calling blinky from my ayaka through golang module. I am going to read blinky packages and implements it on my ayaka.

However, from Readme, blinky does not support another architecture.  Could I ask you how to use it on multi architecture, x86_64 and pentium4 ?

**Comment by BrenekH** [comment]
So the current problem with multi-architecture is that `x86_64` is hardcoded in places like [blinkyd/main.go#L72](https://github.com/BrenekH/blinky/blob/505dc63d64c5fa9b03ffa5d72b164fbc127856d1/cmd/blinkyd/main.go#L72) and [blinkyd/main.go#L81](https://github.com/BrenekH/blinky/blob/505dc63d64c5fa9b03ffa5d72b164fbc127856d1/cmd/blinkyd/main.go#L81). To fix it, each repository would need a list of architectures, somehow defined in the arguments passed via env var or cli. Or maybe the architectures could just be the same for every repo and be configured as a separate argument. It's something I would have to think about and come up with a solution for.

**Comment by Hayao0819** [comment]
They all seem to be used to access architecture-specific directories. I would like to write a function to retrieve those directories. You are right that I need a list of architectures on the source code, but I don't know where I should put that file.I will add the list of architectures and the architectures to be built to the config and repeat for them at build time.

**Comment by Hayao0819** [comment]
I love this project and now start to use it as backend of ayaka.

https://github.com/Hayao0819/ayaka/blob/master/cmd/blinky.go

**Comment by Hayao0819** [comment]
Hey, I noticed that you are working on this issue in multi-arch branch.

And I have started to write `apiunstable/api.go` for handling `any` package.

The current code has all packages in one directory, but any package can share files between architectures.

So I would like to use something like symbolic links. Any opinions on the directory structure for storing files? I will try to write the code according to you.

**Comment by BrenekH** [comment]
I'm not entirely sure what my plan was for `any` packages when I started writing this all out, but I do like the idea of using symlinks.

Ideally for me, the `any` packages go in their own folder and then are symlinked into each architecture-specific directory. After that, the databases (Pacman and Blinky) get updated and then everything _should_ work and be good to go.

**Comment by Hayao0819** [comment]
Is it correct ?

```
any/foo-bar-any.pkg.tar.zst
x86_64/foo-bar-any.pkg.tar.zst (link to any)
```

**Comment by BrenekH** [comment]
Yep! Just remember that any `.sig` files will need to be linked over as well

**Comment by Hayao0819** [comment]
How about the backend structure? How API structure does have info about architecture ?

**Comment by Hayao0819** [comment]
After a long hiatus from Arch Linux I decided to come back and resume this work.

What is currently not implemented is package removal. I am working on this, but am currently experiencing one problem.

At the moment, it was easy to identify the files to be removed by simply storing the file name in the database since it is x86_64 only. However, if we are going to support multiple architectures, we need to know not only the file name, but also the architecture status of the package. Also, since the architecture is not always constant, even if a package has “arch=any”, it may not exist on all architectures.
Due to these issues, I now have to choose between several options for implementation.
- Assume all packages exist on all architectures and attempt to remove them. Ignore the file if it does not exist
- Change the structure of the database to hold two pieces of information for package names: the file name and the architecture name.
- Identify the architecture using string manipulation from the file name and package name 
I will follow the implementation you want to do. How should I do this?

**Comment by BrenekH** [comment]
What makes the most sense to me is to track each architecture independently. That way packages could be deleted from a single architecture and not all of them if required. Obviously the CLI doesn't support this right now, but it could make sense to add a `--arch` flag to the delete command to specify a specific architecture to remove, defaulting to everything.

The database structure I have in my head would be `Key: packageName` maps to `Value: PackageStruct` with the package structure looking like:

```go
type PackageStruct {
    Files map[string]string // Map the architecture slug to the filename
}
```

Names and what not can be different, but this structure allows architectures to be mapped to individual files and it's all nicely encapsulated.

**Comment by Hayao0819** [comment]
How do you map to the KV Store? Only []bytes are supported and must be mapped via some format such as JSON.

Not directly related to the topic, but #27 is in working order, please review. There are a few redundancies, but I couldn't think of a specific fix.

**Comment by BrenekH** [comment]
Serializing into JSON would be fine but it could also be serialized into a binary format with something like [encoding/gob](https://pkg.go.dev/encoding/gob). I'm not entirely sure performance wise if gob is better than JSON, but my gut feeling is that that is the case. It's really up to you though.

**Comment by BrenekH** [comment]
I agree that it should be an internal type to `keyvaluestore`. Methods like `StorePackageFile` will also need to be updated to take an arch parameter so the adapter knows which filename the API code is operating on.

**Comment by Hayao0819** [comment]
I wrote #34 using Gob just to get the information.
If there are no style issues, directional differences, etc. with this code, I will also implement deletions and additions.

**Comment by BrenekH** [comment]
> If there are no style issues, directional differences, etc. with this code, I will also implement deletions and additions.

Looks all good to me!
